### PR TITLE
Add pixelset list view filtering

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Add filters to ease pixel set search.

## Proposal

- [x] Filter `PixelSet` given
  * species
  * omics unit types
  * omics areas
  * tags
- [x] Integrate the filter form
- [x] Improve pagination handling along with filters
- [x] Battle test filters

### CODE REVIEW

**Comment by jmaupetit** [comment]
Current status:

screenshot-2018-1-23 pixel sets

We need to work on analysis and experiment descriptions that are too long to be displayed in a long table.

**Comment by willdurand** [comment]
> We need to work on analysis and experiment descriptions that are too long to be displayed in a long table.

Can't we have horitzontal tabs maybe? 2 columns (title/value)

File: Pipfile.lock [hunk 1]
```diff
@@ -66,10 +66,10 @@
         },
         "django-js-asset": {
             "hashes": [
-                "sha256:0dd2c5f64f2b24eb8a7270a6a59cb914a03f205335bd0eb6207bf61cf7410828",
-                "sha256:c6637e4f3fef12024564057a48bef484db6491c610273d76d25d8dbe6ecc4ac1"
+                "sha256:bcd00d7dc67f86cad1dac4bf7b43fa4b75088c214d2c92002b8705c6a7362e7e",
+                "sha256:38fc1de7a93a2f667fe7ee34acc53d682bac5ce6c73ac6f0ecb4ff77d0f61f1b"
             ],
-            "version": "==0.1.1"
+            "version": "==1.0.0"
```
Comment (willdurand) [inline Pipfile.lock]: That's the problem of not adding constraints to the `Pipfile`. I would expect changes and BC breaks between `0.1.1` and `1.0.0`. Can we do something for production-related packages?

File: apps/explorer/tests/test_views.py [hunk 2]
```diff
@@ -50,3 +50,230 @@ def test_renders_pixelset_list(self):
...
+        make_development_fixtures(
+            n_pixel_sets=n_pixel_sets,
+            n_pixels_per_set=1
+        )
+
+        # no filter
+        response = self.client.get(self.url)
+        self.assertContains(
+            response,
+            '<tr class="pixelset">',
```
Comment (willdurand) [inline apps/explorer/tests/test_views.py]: This seems a bit weak. Can't you pass a CSS element instead?

File: apps/explorer/tests/test_views.py [hunk 3]
```diff
@@ -50,3 +50,230 @@ def test_renders_pixelset_list(self):
...
+        # omics_unit_types filter
+        omics_unit_type = models.OmicsUnitType.objects.first()
+        data = {
+            'omics_unit_types': [omics_unit_type.id, ]
+        }
+        response = self.client.get(self.url, data)
+        self.assertContains(
+            response,
+            '<tr class="pixelset">',
+            count=int(n_pixel_sets / 2)
```
Comment (willdurand) [inline apps/explorer/tests/test_views.py]: Isnt it a `int` by default in Python 3?

File: assets/scss/_explorer.scss [hunk 4]
```diff
@@ -1,9 +1,48 @@
...
+          label {
+            font-weight: bold;
+
+            ul {
+              max-height: 200px;
+              margin-left: 0;
+              list-style: none;
+              overflow-y: auto;
+
+              &#id_tags {
```
Comment (willdurand) [inline assets/scss/_explorer.scss]: mmh, it is an id here? why do you need to nest it? it should be unique and used once.

File: assets/scss/_explorer.scss [hunk 5]
```diff
@@ -1,9 +1,48 @@
 .explorer {
 
-  &.pixelsets {
+  aside.sidebar-left {
+    @include xy-cell($size: 2.5, $gutter-output: false);
+
+    .filters {
+
+      form {
```
Comment (willdurand) [inline assets/scss/_explorer.scss]: we could have a class for that, and avoid HTML elements as much as possible.

File: apps/explorer/tests/test_views.py [hunk 6]
```diff
@@ -50,3 +50,230 @@ def test_renders_pixelset_list(self):
...
+            'omics_unit_types': [omics_unit_type.id, ],
+            'omics_area': [omics_area.id, ],
+            'tags': [tag.id, ],
+        }
+        response = self.client.get(self.url, data)
+        self.assertContains(
+            response,
+            '<tr class="pixelset">',
+            count=1
+        )
```
Comment (willdurand) [inline apps/explorer/tests/test_views.py]: We need tests when things go wrong, like no results and bad `GET` parameter given.

File: apps/explorer/tests/test_views.py (hunk 2 above)
Comment (jmaupetit) [reply apps/explorer/tests/test_views.py]: Do you mean a CSS selector like `tr.pixelset`?

File: apps/explorer/tests/test_views.py (hunk 3 above)
Comment (jmaupetit) [reply apps/explorer/tests/test_views.py]: Yes my lord. Old habits.

File: apps/explorer/tests/test_views.py (hunk 6 above)
Comment (jmaupetit) [reply apps/explorer/tests/test_views.py]: Right.

File: assets/scss/_explorer.scss (hunk 4 above)
Comment (jmaupetit) [reply assets/scss/_explorer.scss]: Oops. My bad. That's silly.

File: assets/scss/_explorer.scss (hunk 5 above)
Comment (jmaupetit) [reply assets/scss/_explorer.scss]: Yup. Sounds safer.

File: Pipfile.lock (hunk 1 above)
Comment (jmaupetit) [reply Pipfile.lock]: I understand your concern. That's a second degree dependency. There is something I still don't understand in the `Pipfile.lock`'s generation/update: when installing a new dependency _via_: `pipenv install foo`, `pipenv` will automatically upgrade all dependencies. Most of the time you do not want that. Dependencies should be explicitly upgraded, _i.e._ when I ask for it.

Comment (willdurand) [reply Pipfile.lock]: Yes, and there is an issue for that on Pipenv. Apparently, it wont get solved.. :(

File: apps/explorer/tests/test_views.py (hunk 2 above)
Comment (willdurand) [reply apps/explorer/tests/test_views.py]: Or only `.pixelset`

File: assets/scss/_explorer.scss (hunk 4 above)
Comment (willdurand) [reply assets/scss/_explorer.scss]: no problemo

File: apps/explorer/tests/test_views.py (hunk 2 above)
Comment (jmaupetit) [reply apps/explorer/tests/test_views.py]: Unfortunately no. But, note that the assertion is in "HTML mode", it's a little bit weaker. See the [documentation](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.SimpleTestCase.assertContains).

File: apps/explorer/tests/test_views.py (hunk 3 above)
Comment (jmaupetit) [reply apps/explorer/tests/test_views.py]: Meh… answered too quickly. In fact that's the opposite: this is the case in python 2, python 3 always returns a `float` as a division result.

Comment (willdurand) [reply apps/explorer/tests/test_views.py]: ah, then is it... `//` that we should be using now?

File: apps/explorer/tests/test_views.py (hunk 2 above)
Comment (jmaupetit) [reply apps/explorer/tests/test_views.py]: Note that the assertion **should** be in `html=True` mode. I'll fix this.

**Comment by jmaupetit** [comment]
> Can't we have horitzontal tabs maybe? 2 columns (title/value)

Yes, I'll try that in a new PR.

**Comment by jmaupetit** [comment]
> One Scss file is too nested

The `_explorer.scss`?

**Comment by willdurand** [comment]
> The _explorer.scss?

yes

**Comment by jmaupetit** [comment]
I did my best to decrease what I can decrease. It should be ok now.

**Pull request merged**